### PR TITLE
Smart fridge

### PR DESCRIPTION
**Pull request by Fishfish458**
## About the PR 

Added basic functionality for smart fridge

Checks for tags: Pill Produce Bottle to store any of those types
on use with any bag will check the top layer for anything that might fit into that criteria and add it to the smart fridge

**Screenshots**

adds pills
adds produce
ejecting

**Changelog**

:cl:
- add: Add functionality to the Smart Fridges

### CODE REVIEW

File: Content.Client/SmartFridge/UI/SmartFridgeMenu.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            }
+
+            SetSize = ((250), (350));
+        }
+
+        public void ItemSelected(ItemList.ItemListSelectedEventArgs args)
+        {
+            // attempt eject
+            Owner.Eject(_cachedInventory[args.ItemIndex].ID);
+        }
```
Comment (metalgearsloth) [inline Content.Client/SmartFridge/UI/SmartFridgeMenu.xaml.cs:56]: I smell copy-paste.

This should ideally be generic with vending machines if their functionality is intended to be similar.

File: Content.Client/SmartFridge/SmartFridgeBoundUserInterface.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            base.Dispose(disposing);
+            if (!disposing)
+                return;
+
+            _menu?.Dispose();
+        }
```
Comment (metalgearsloth) [inline Content.Client/SmartFridge/SmartFridgeBoundUserInterface.cs:55]: Here too looks like copy-paste.

File: Content.Server/SmartFridge/SmartFridgeComponent.cs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        public SoundSpecifier SoundVend = new SoundPathSpecifier("/Audio/Machines/machine_vend.ogg");
+        [DataField("soundDeny")]
+        // Yoinked from: https://github.com/discordia-space/CEV-Eris/blob/35bbad6764b14e15c03a816e3e89aa1751660ba9/sound/machines/Custom_deny.ogg
+        public SoundSpecifier SoundDeny = new SoundPathSpecifier("/Audio/Machines/custom_deny.ogg");
+        [ViewVariables] public BoundUserInterface? UserInterface => Owner.GetUIOrNull(SmartFridgeUiKey.Key);
+
+        [DataField("whitelist")]
+        public EntityWhitelist? Whitelist;
+        public Container? Storage = default!;
+        public Dictionary<uint,  Queue<EntityUid>> entityReference = new();
```
Comment (metalgearsloth) [inline Content.Server/SmartFridge/SmartFridgeComponent.cs:30]: You cooked some of the copy-paste here.

I'm not sure if there should be a shared comp for smartfridge / vending machines but atm I'm leaning towards yes based on this.

File: Content.Server/SmartFridge/SmartFridgeSystem.cs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+            else if (!IsPowered(uid, fridgeComponent))
+            {
+                finalState = SmartFridgeVisualState.Off;
+            }
+
+            if (TryComp<AppearanceComponent>(fridgeComponent.Owner, out var appearance))
+            {
+                appearance.SetData(SmartFridgeVisuals.VisualState, finalState);
+            }
+        }
```
Comment (metalgearsloth) [inline Content.Server/SmartFridge/SmartFridgeSystem.cs:265]: Wayyy too much copy-paste here.

File: Content.Shared/SmartFridge/SharedSmartFridgeComponent.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
+using Robust.Shared.Serialization;
+
+namespace Content.Shared.SmartFridge
+{
+    public class SharedSmartFridgeComponent : Component
```
Comment (metalgearsloth) [inline Content.Shared/SmartFridge/SharedSmartFridgeComponent.cs:5]: Classes should never just be virtual now (this applies to all in the PR) and your IDE is probably giving you warnings as such.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Needs trimming down a lot because several hundred lines is just copy-pasted from vending machines.

**Comment by Fishfish458** [comment]
I guess I really don't understand what i'm doing with this, it's gonna take a bit longer to fix so will keep the same pr but open again later

**Pull request closed without merging**